### PR TITLE
Document the S3Queue/AzureQueue power-loss data-loss window and its fsync mitigation

### LINKED ISSUE #111537: S3Queue silently loses consumed rows on power loss (at-least-once becomes at-most-once; unrecoverable with after_processing='delete')
### Describe what's wrong

`S3Queue` (`StorageObjectStorageQueue`) can **silently and permanently lose already-consumed rows** after a power loss of the ClickHouse node, converting its documented at-least-once delivery into at-most-once. With `after_processing = 'delete'` the loss is **unrecoverable** because the source object is deleted before the ingested rows are durable.

The root cause is a cross-domain ordering gap in `StorageObjectStorageQueue::commit`:

1. `streamToViews` runs the insert pipeline, so the consumed rows land in the target `MergeTree` part — in the **page cache only** (no fsync; `fsync_after_insert` is 0 by default and there is no `S3Queue` setting to order durability before the commit). `src/Storages/ObjectStorageQueue/StorageObjectStorageQueue.cpp:934-935,980`
2. Inside `commit`, when `after_processing != KEEP`, the source objects are **deleted** (`removeObjectsIfExist`). `StorageObjectStorageQueue.cpp:1049-1053` → `src/Storages/ObjectStorageQueue/ObjectStorageQueuePostProcessor.cpp:75`
3. Then the "processed" nodes are committed to **Keeper** via `tryMulti`. `StorageObjectStorageQueue.cpp:1078-1079`

Keeper's commit is force-synced and, being a separate service (or a separate node), is unaffected by *this* node's power loss. So if the ClickHouse node loses power after step 3 but before the target part is fsynced, on restart Keeper reports the files as already processed, `S3Queue` never re-reads them, and the rows are gone. This is a device-level power loss (dirty page-cache loss); a plain `SIGKILL` does **not** expose it because the page cache survives a process kill — which is why normal crash testing misses it.

This contradicts the documented behavior: the "Limitations" section of the `S3Queue` docs lists "abnormal server termination" **only** as a cause of *duplicated* rows, never loss (`docs/en/engines/table-engines/integrations/s3queue.md`, "Limitations").

The same ordering exists in the `Kafka` engine (broker offset committed after an un-fsynced insert) and `FileLog` (un-fsynced `.meta` offset marker), but `S3Queue` is the clearest case because Keeper is a genuinely independent durability domain and `after_processing = 'delete'` destroys the only other copy.

### Does it reproduce on the most recent release?

Yes — reproduced on `26.4.1.1` and on a recent `master` build `26.7.1.1258` (identical result). The code path is present on current `master`.

### How to reproduce

Topology: only the ClickHouse node loses power; Keeper and the object store stay alive (the realistic case — Keeper/S3 are separate services).

- ClickHouse server `<path>` on a filesystem backed by a `dm-flakey` device (faithful power loss = drop all un-fsynced writes).
- Keeper as a separate process on a normal filesystem (survives).
- MinIO on a normal filesystem (survives).

Steps:

1. `CREATE TABLE tgt (id UInt64, val UInt64) ENGINE = MergeTree ORDER BY id;` (default settings, `fsync_after_insert = 0`)
2. `CREATE TABLE q (id UInt64, val UInt64) ENGINE = S3Queue('http://minio/bkt/*.csv', 'user', 'pass', 'CSV') SETTINGS mode = 'unordered', keeper_path = '/s3q', after_processing = 'delete';`
3. `CREATE MATERIALIZED VIEW mv TO tgt AS SELECT id, val FROM q;`
4. Upload N files to the bucket. Wait until all rows are ingested into `tgt` and the processed nodes appear under `keeper_path/processed`.
5. Cut power to the ClickHouse node (drop all un-fsynced writes) — Keeper and MinIO stay up.
6. Restart the ClickHouse node.

Result: `SELECT count() FROM tgt` is 0 (all consumed rows lost); the processed nodes still exist in Keeper so `S3Queue` never re-reads the files; with `after_processing = 'delete'` the source objects are already gone → unrecoverable.

A self-contained harness (MinIO + external Keeper + `dm-flakey`, with a fsync-off "canary" table that proves the power-loss window is live, and a matched `fsync_after_insert = 1` control) is available and reproduces this deterministically:

<details><summary>Measured results (20 files × 50 rows = 1000 consumed rows)</summary>

| scenario | target `fsync_after_insert` | outcome |
|---|---|---|
| graceful stop/restart | 0 | 1000/1000 survive (control passes) |
| **power loss (default)** | **0** | **tgt 1000 → 0; Keeper keeps all 20 processed markers; NOT re-ingested; canary lost → window live** |
| power loss | 1 | 1000/1000 survive (part fsynced during `execute()` before the Keeper commit) |

Identical shape with `after_processing = 'delete'`, except the loss is unrecoverable (source deleted at `commit` before the cut).
</details>

### Expected behavior

An acknowledged/consumed file's rows should be durable before `S3Queue` records the file as processed (and, for `after_processing = 'delete'`, before the source object is removed) — i.e. the engine should honor at-least-once across a power loss, or expose a setting to order local durability ahead of the processed-commit. At minimum the loss window should be documented alongside the existing duplicate-only limitation.

### Error message and/or stacktrace

No error — the loss is silent. After recovery `SELECT count() FROM tgt` is simply short, and `system.zookeeper` under `keeper_path/processed` still lists the lost files.

### Additional context

Setting `fsync_after_insert = 1` (and `fsync_part_directory = 1`) on the target table closes the direct-to-table window, but that is off by default, is not documented as required for `S3Queue` correctness, and does not help an operator who relies on the documented at-least-once guarantee. Found by device-level power-loss testing (dm-flakey); invisible to `SIGKILL`-based crash tests because the page cache survives a process kill.

### Version info
- Resolved by: #111571
- Merged into: `26.8.1.90` (included in `26.8` and later)

### PR DESCRIPTION
**Pull request by groeneai**
### Changelog category (leave one):
- Documentation (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
...

### Description

Closes: https://github.com/ClickHouse/ClickHouse/issues/111537

The `S3Queue`/`AzureQueue` "Limitations" docs listed "abnormal server termination" only as a cause of *duplicated* rows. As reported in #111537, a device-level power loss of the ClickHouse node can also silently *lose* consumed rows: `StorageObjectStorageQueue::commit` records a file as processed in Keeper (and, with `after_processing = 'delete'`, removes the source object) as soon as the insert pipeline finishes, but the rows inserted into the target `MergeTree` table are only durable once the part is fsynced, which does not happen synchronously by default (`fsync_after_insert = 0`). Keeper is force-synced and usually runs on a separate node, so it survives this node's power loss; if the node loses power after the file is committed as processed but before the target part is fsynced, the file is not re-read on restart and its rows are gone (unrecoverable with `after_processing = 'delete'`). A plain process kill does not expose this because the page cache survives it.

This documents the loss window and the operator mitigation the reporter measured (`fsync_after_insert = 1` and `fsync_part_directory = 1` on the target table), scoped to the recommended materialized-view consumption path where the file is committed only after the whole insert pipeline finishes. It updates all four doc copies: the standalone `s3queue.md` and `azure-queue.md` pages and the embedded engine documentation in `registerQueueStorage.cpp` (surfaced via `system.table_engines.description`). `AzureQueue` is covered because it uses the identical `StorageObjectStorageQueue` implementation.

Documentation only, no behavior change. A code-level fix that lets the engine force target durability before the processed-commit would be a cross-engine durability-contract change (the same ordering exists in `Kafka` and `FileLog`) with backward-compatibility and default-behavior implications; that design decision is raised separately on the issue for the component owner.

### Version info
- Merged into: `26.8.1.90` (included in `26.8` and later)

### CODE REVIEW

**Comment by groeneai** [comment]
<details>
<summary><b>Internal second-model review</b> (click to expand)</summary>

An independent second-model review was run against this change. Findings and dispositions:

**❌ Blockers**

- **Mitigation over-claimed for the direct-select path** (AGREE, fixed). For background materialized-view consumption `StorageObjectStorageQueue::commit` runs only after the whole insert pipeline finishes, so `fsync_after_insert=1` orders part durability before the processed-commit. For direct `INSERT ... SELECT` with `commit_on_select=1` the source commits at its own EOF before the destination sink finalizes its last part, so the mitigation does not reliably apply there. The docs now scope the mitigation to the materialized-view path and explicitly note the direct-select exception.

**⚠️ Majors**

- **Ask for device-level power-loss evidence** (partial AGREE). The mechanism (MV-path sink finalization ordered before the Keeper commit) is confirmed by reading the source, and the issue reporter already provides a matched `dm-flakey` control (`fsync_after_insert=0` loses rows, `=1` preserves them). Rather than re-run a hardware power-cut for a documentation change, the wording was softened from "closes this window" to "narrows this window substantially" in all four copies.

**💡 Nits**

- **Use the full issue URL in the relationship line** (AGREE, fixed). The PR body uses `Closes: https://github.com/ClickHouse/ClickHouse/issues/111537`.

</details>

Session id: cron:clickhouse-author-slot-2:20260723-083400

**Comment by groeneai** [comment]
<details>
<summary><b>Pre-PR validation gate</b> (click to expand)</summary>

| # | Question | Answer |
|---|---|---|
| a | Deterministic repro? | Reporter's `dm-flakey` device-power-loss harness in #111537 deterministically loses rows (`fsync_after_insert=0` -> 0/1000 survive; `=1` -> 1000/1000). Documentation change; no new repro required. |
| b | Root cause explained? | Yes. `StorageObjectStorageQueue::commit` records the file as processed in Keeper (force-synced, independent domain) and, with `after_processing='delete'`, deletes the source, before the target `MergeTree` part is fsynced (`fsync_after_insert=0` by default, part in page cache). Power loss between the Keeper commit and the part fsync loses the rows. |
| c | Fix matches root cause? | Yes for this PR's scope: it documents the exact loss window and the verified mitigation. The code-level durability contract is deliberately deferred to the issue (cross-engine, backward-compat, owner decision). |
| d | Test intent preserved / new tests? | N/A - documentation only. |
| e | Demonstrated both directions? | N/A (docs). Rendered markdown verified; embedded `R"DOCS_MD(...)"` string delimiters remain balanced. |
| f | General, not a narrow patch? | Yes. Covers both `S3Queue` and `AzureQueue` (identical `StorageObjectStorageQueue` implementation) and both doc copies each (standalone `.md` + embedded doc surfaced via `system.table_engines.description`). |
| g | Generalizes across inputs/types? | N/A (docs). |
| h | Backward compatible? | Yes - documentation only, zero behavior/setting/format change. |
| i | Invariants/contracts preserved? | Yes - no code path touched; the `.cpp` edits are inside a raw string literal only. |

</details>

Session id: cron:clickhouse-author-slot-2:20260723-083400

**Comment by groeneai** [comment]
cc @kssenii - could you review this? Docs-only: documents the S3Queue/AzureQueue power-loss data-loss window in `StorageObjectStorageQueue::commit` (file committed to Keeper + source deleted before the target part is fsynced) and the `fsync_after_insert=1` mitigation, scoped to the materialized-view path. Closes #111537.

File: docs/en/engines/table-engines/integrations/azure-queue.md [hunk 1]
```diff
@@ -195,3 +195,7 @@ exception:
 1 row in set. Elapsed: 0.002 sec.
 
 ```
+
+## Limitations {#limitations}
+
+`AzureQueue` shares the same implementation as `S3Queue` and has the same [limitations](/engines/table-engines/integrations/s3queue#limitations). In particular, a device-level power loss of the ClickHouse node can silently lose consumed rows: a file is recorded as processed in Keeper (and, with `after_processing = 'delete'`, its source blob removed) as soon as the insert finishes, but the inserted rows are only durable once the target part is fsynced, which does not happen synchronously by default (`fsync_after_insert = 0`). For the recommended materialized-view consumption path, setting `fsync_after_insert = 1` (and `fsync_part_directory = 1`) on the target `MergeTree` table narrows this window substantially.
```
Comment (kssenii) [inline docs/en/engines/table-engines/integrations/azure-queue.md:201]: This is not a property of S3(Azure)Queue, but what is the underlying destination table and insert/MV settings. So I do not see why we need to put this into s3 queue documentation. The same behaviour have all the MV-based table engines.

Comment (groeneai) [reply docs/en/engines/table-engines/integrations/azure-queue.md:201]: Agreed that the un-fsynced-insert window is generic: it is a property of the target table and `fsync_after_insert`, and every MV-based engine (Kafka, FileLog, a plain source -> MV chain) shares it.

What is S3Queue/AzureQueue-specific is `after_processing = 'delete'`: the source object is removed in the same commit, so after the loss there is no surviving copy to rebuild from. A generic MV chain keeps its source data and can be replayed; here the loss is permanent. That asymmetry is what #111537 asks us to surface.

I will narrow these docs to just that engine-specific point (unrecoverable loss under `after_processing = 'delete'`) and drop the generic-durability prose, unless you would rather it live in a generic insert-durability page with a one-line pointer from here.

**Comment by groeneai** [comment]
### CI finish ledger - 3a33557f78fd0da5b2eb08d682d36d6c929093c3

CI fully finished on this head. This PR is documentation-only (s3queue/azure-queue `.md` plus the embedded help-string text in `registerQueueStorage.cpp`), so neither failure below can be PR-caused. Every failure has an owner.

| Check / test | Reason | Owner |
|---|---|---|
| Stateless (amd_llvm_coverage, old analyzer, s3 storage, DBReplicated, WasmEdge, parallel, 2/3) / Logical error `!part.empty()` (+ "Server died") | old-analyzer query-tree to AST error-formatting LOGICAL_ERROR when a dotted lambda-argument name is split on `.` (aborts under coverage `abort_on_logical_error`). Chronic trunk: 2 master + 44+ unrelated PRs, 30d. Unrelated to a docs change. | a fix task is being worked at full effort (follow-up to merged #108735 which fixed only the LambdaNode path; a second ASTIdentifier split path remains). Fixing PR link to follow here. |
| Stateless (amd_asan_ubsan, distributed plan, parallel) / 02481_i43247_ubsan_in_minmaxany | flaky Code 241 MEMORY_LIMIT_EXCEEDED (~55% rerun-fail under the sanitizer distributed-plan config); 2 master + 19 unrelated PRs, 30d. Belongs to the documented amd_asan_ubsan distributed-plan runner memory-pressure family. Unrelated to a docs change. | a fix task is moved to pending (full effort). Fixing PR link to follow here. |
| Mergeable Check / PR | rollup of the above | owned via the two lines above |
| CH Inc sync | - | CH Inc sync (private, not actionable) |

Session id: cron:our-pr-ci-monitor:20260723-170000

**Review by kssenii — APPROVED** [review]
